### PR TITLE
Test refactor

### PR DESCRIPTION
**Pull request by Abrams27**
I changed a bit test and added partial results

### CODE REVIEW

File: third_party.bzl [hunk 1]
```diff
@@ -23,6 +23,10 @@ _deps = [
     _dependency("org.eclipse.xtext:org.eclipse.xtext.xbase.lib:2.19.0"),
     _dependency("commons-cli:commons-cli:jar:1.4"),
     _dependency("org.jetbrains:annotations:jar:20.1.0"),
+    _dependency("io.vavr:vavr:0.9.0"),
+    _dependency("org.projectlombok:lombok:1.18.4"),
+    _dependency("org.apache.logging.log4j:log4j-api:2.6.1"),
+    _dependency("org.apache.logging.log4j:log4j-core:2.6.1"),
```
Comment (gerardd33) [inline third_party.bzl:29]: Why did you add Lombok and a separate logging tool and not Lombok's Slf4j? Is Lombok used for something else?

**Review by gerardd33 — APPROVED** [review]

File: third_party.bzl (hunk 1 above)
Comment (Abrams27) [reply third_party.bzl:29]: I forgot to delete this dep after my experiments

File: test.sh [hunk 2]
```diff
@@ -1,16 +1,57 @@
...
 
+
+log_test_progress "Running BSP tests..."
+echo -e "===================================\n"
+
 run_test test_bsp_server
+
+echo -e "\n\n==================================="
+echo -e "==================================="
+log_test_progress "${GREEN}All test passed!"
```
Comment (agluszak) [inline test.sh:57]: Shouldn't this part be shown only when all tests actually passed?

**Review by agluszak — COMMENTED** [review]
I would consider incorporating Bazel's `sh_test` in order to make this test possible to run via `bazel test`. Also the test running logic is complicated enough that IMO it deserves a unit test of its own... Otherwise LGTM.

**Review by agluszak — CHANGES_REQUESTED** [review]
I would consider incorporating Bazel's `sh_test` in order to make this test possible to run via `bazel test`. Also the test running logic is complicated enough that IMO it deserves a unit test of its own...

Although this PR runs green on the CI, it looks like it's a false positive - probably introduces by the changes themselves
```
21:51:13.136 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerSingleTest - Running "resolve project" test...
Exception in thread "Thread-1" java.util.concurrent.CancellationException
	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
	at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-4" java.util.concurrent.CancellationException
	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
	at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-5" java.util.concurrent.CancellationException
	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
	at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-2" java.util.concurrent.CancellationException
21:59:58.598 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerSingleTest - Test "resolve project" passed!
	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
	at java.lang.Thread.run(Thread.java:748)
```

and then

```
[TEST] Test "test_bsp_server" successful (846 sec) 


===================================
===================================

[TEST] All test passed!
```

Also the logging format is inconsistent between the logs from our java code:
```
21:51:12.715 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerTestRunner - Starting BazelBspServerTest...
```
 Bazel:
```
Nov 26, 2020 9:51:07 PM rules.jvm.external.maven.MavenPublisher lambda$writeFile$2
INFO: Copying /tmp/artifact.jar14931761111460028011.sha1 to file:///home/runner/.m2/repository/org/jetbrains/bsp/bazel-bsp/0.1.0/bazel-bsp-0.1.0.jar.sha1
```
and bash test runner:
```
[TEST] Installing BSP...
```

File: test.sh (hunk 2 above)
Comment (Abrams27) [reply test.sh:57]: but it is shown only when all tests passed:
```
log_test_progress "${RED}Test \"$TEST_ARG\" failed $NC ($DURATION sec) $NC"
exit $EXECUTION_CODE
```

if any test fails script exit

**Comment by Abrams27** [comment]
> I would consider incorporating Bazel's `sh_test` in order to make this test possible to run via `bazel test`.

great idea - I'll add 

> Also the test running logic is complicated enough that IMO it deserves a unit test of its own... 

hmm, I think that logic is simple and linear - imo crucial part is creating testing data, but it cannot be tested

> Although this PR runs green on the CI, it looks like it's a false positive - probably introduces by the changes themselves
> 
> ```
> 21:51:13.136 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerSingleTest - Running "resolve project" test...
> Exception in thread "Thread-1" java.util.concurrent.CancellationException
> 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> 	at java.lang.Thread.run(Thread.java:748)
> Exception in thread "Thread-4" java.util.concurrent.CancellationException
> 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> 	at java.lang.Thread.run(Thread.java:748)
> Exception in thread "Thread-5" java.util.concurrent.CancellationException
> 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> 	at java.lang.Thread.run(Thread.java:748)
> Exception in thread "Thread-2" java.util.concurrent.CancellationException
> 21:59:58.598 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerSingleTest - Test "resolve project" passed!
> 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> 
> and then
> 
> ```
> [TEST] Test "test_bsp_server" successful (846 sec) 
> 
> 
> ===================================
> ===================================
> 
> [TEST] All test passed!
> ```

I will try to investigate it (any ideas how to reproduce it locally?)

> Also the logging format is inconsistent between the logs from our java code:
> 
> ```
> 21:51:12.715 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerTestRunner - Starting BazelBspServerTest...
> ```
> 
> Bazel:
> 
> ```
> Nov 26, 2020 9:51:07 PM rules.jvm.external.maven.MavenPublisher lambda$writeFile$2
> INFO: Copying /tmp/artifact.jar14931761111460028011.sha1 to file:///home/runner/.m2/repository/org/jetbrains/bsp/bazel-bsp/0.1.0/bazel-bsp-0.1.0.jar.sha1
> ```
> 
> and bash test runner:
> 
> ```
> [TEST] Installing BSP...
> ```

any more consistent format proposition? (can I change bazel's logs format?)

**Comment by Abrams27** [comment]
Final conclusion

> > I would consider incorporating Bazel's `sh_test` in order to make this test possible to run via `bazel test`.
> 
> great idea - I'll add
> 

Great idea, but... the tricky one. I will do it in a separate PR later.


> > Also the test running logic is complicated enough that IMO it deserves a unit test of its own...
> 
> hmm, I think that logic is simple and linear - imo crucial part is creating testing data, but it cannot be tested
> 
> > Although this PR runs green on the CI, it looks like it's a false positive - probably introduces by the changes themselves
> > ```
> > 21:51:13.136 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerSingleTest - Running "resolve project" test...
> > Exception in thread "Thread-1" java.util.concurrent.CancellationException
> > 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> > 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> > 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> > 	at java.lang.Thread.run(Thread.java:748)
> > Exception in thread "Thread-4" java.util.concurrent.CancellationException
> > 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> > 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> > 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> > 	at java.lang.Thread.run(Thread.java:748)
> > Exception in thread "Thread-5" java.util.concurrent.CancellationException
> > 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> > 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> > 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> > 	at java.lang.Thread.run(Thread.java:748)
> > Exception in thread "Thread-2" java.util.concurrent.CancellationException
> > 21:59:58.598 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerSingleTest - Test "resolve project" passed!
> > 	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
> > 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:54)
> > 	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor$1.get(ConcurrentMessageProcessor.java:50)
> > 	at ch.epfl.scala.bsp.testkit.client.mock.MockSession.$anonfun$startServerConnection$1(MockSession.scala:31)
> > 	at java.lang.Thread.run(Thread.java:748)
> > ```
> > 
> > 
> > and then
> > ```
> > [TEST] Test "test_bsp_server" successful (846 sec) 
> > 
> > 
> > ===================================
> > ===================================
> > 
> > [TEST] All test passed!
> > ```
> 
> I will try to investigate it (any ideas how to reproduce it locally?)
> 

I found out that the problem is in the test client - I will create an issue with it. It means that I didn't break anything.

> > Also the logging format is inconsistent between the logs from our java code:
> > ```
> > 21:51:12.715 [main] INFO  org.jetbrains.bsp.bazel.BazelBspServerTestRunner - Starting BazelBspServerTest...
> > ```
> > 
> > 
> > Bazel:
> > ```
> > Nov 26, 2020 9:51:07 PM rules.jvm.external.maven.MavenPublisher lambda$writeFile$2
> > INFO: Copying /tmp/artifact.jar14931761111460028011.sha1 to file:///home/runner/.m2/repository/org/jetbrains/bsp/bazel-bsp/0.1.0/bazel-bsp-0.1.0.jar.sha1
> > ```
> > 
> > 
> > and bash test runner:
> > ```
> > [TEST] Installing BSP...
> > ```
> 
> any more consistent format proposition? (can I change bazel's logs format?)

It will be tough too, so I will leave it as it is now, maybe I will try to fix it later

**Comment by andrefmrocha** [comment]
> I would consider incorporating Bazel's `sh_test` in order to make this test possible to run via `bazel test`
Based on what we discussed last week, I'm not exactly sure if this is possible but the [bazel](https://github.com/bazelbuild/bazel/) itself might be a good starting point for this

**Comment by andrefmrocha** [comment]
> I found out that the problem is in the test client - I will create an issue with it. It means that I didn't break anything.
This probably is an exception that is being printed and shouldn't. All should be fine on your side

**Comment by Abrams27** [comment]
> > I found out that the problem is in the test client - I will create an issue with it. It means that I didn't break anything.
> 
> This probably is an exception that is being printed and shouldn't. All should be fine on your side

@andrefmrocha Could you elaborate your thought?

**Comment by andrefmrocha** [comment]
> > > I found out that the problem is in the test client - I will create an issue with it. It means that I didn't break anything.
> > 
> > This probably is an exception that is being printed and shouldn't. All should be fine on your side
> 
> @andrefmrocha Could you elaborate your thought?

It means that there is a "bug" on the test client side of the BSP that I should look into in the future. It shouldn't be anything related to the tests themselves.
Since this was not outputted beforehand,  it was hidden from us. In any case, it doesn't affect its behavior

**Review by agluszak — APPROVED** [review]
